### DISCUSSION TITLE
CR+LF line terminators question

### DISCUSSION THREAD

**Discussion by feddar** [joomengine/Joomla-Component-Builder#638] [Q&A] [answered]
Hi, Joomla Component Builder create files with "DOS" CR+LF. I'm using Linux and when I edit files with Vim I see every line ending with ^M characters.
Is there a way to tell JCB to not create "DOS" files but "Unix" files with only LF character? Thanks!

**Comment by Llewellynvdm** [comment]
JCB use the PHP global line break `PHP_EOL` constant, so it builds based on your system. Hmmm when you compile on Linux the line ending should be correct.

I can move the line endings to a global value I suppose, like I did with the [tabs](https://github.com/vdm-io/Joomla-Component-Builder/commit/f7be7ddad). They are all over the place, and updating them will be a serious task, but I can look at that option.

I am working on a Linux/Ubuntu 20 system, and so JCB has only been maintained from the Unix base.

Yet building your component on a DOS system will still effect the file type being created since we are using Joomla API to create the file.... hmmm that will be problematic. I mean changing the line breaks and yet the wrong file type will take us back the original problem and why in the first place moved to `PHP_EOL` away from a hard coded option.

I will need to read over the code of the File class in Joomla to see if there is an override option.

**Comment by Llewellynvdm** [comment, marked as answer]
@feddar I am looking over all the discussions that has not resulted in a conclusion... just with a quick browse.

So I read you question again, and my response... seems like I missed the fact that you are compiling on Linux.

Here is what I get...
![image](https://user-images.githubusercontent.com/5607939/110058945-f9bfc780-7d6b-11eb-9bec-0378cfe24861.png)

So from what I can see, there are part in this document with the ^M and parts that do not have it. So I found that this must have happened at some point when a contributor touched the template files (the templates are not build again, but are static)... which is a real pain. I am going to work over them all to make sure they are making use of the correct line brakes. So the dynamic generated code is done correctly, it is an issue with the static files.

**Reply by Llewellynvdm** [reply]
hmmm that also seem to be false... lets try again since here is the template file, and it is not having the M^

![image](https://user-images.githubusercontent.com/5607939/110059481-fbd65600-7d6c-11eb-84b4-bce7110c7d20.png)

So back to square one... I have no idea how this is happening here, do you have any ideas?

We have the correct line breaks in the template, we have the correct line breaks in the dynamic code being build. Yet here we see this M^ jump back up.

So my idea is that the file may have originated in MS... from Joomla, and then as I edited the files it was saved as Unix based files, but not completely... and so when the compiler takes the file and move and merge stuff php picks up this remnants and then warps the line breaks... so my first try is to [dos2unix](https://unix.stackexchange.com/a/32504/130056) all the template files, and see what we get.

Let me know if you have any other ideas....

**Reply by Llewellynvdm** [reply]
Well seems I was right.. problem solved, dos2unix worked... and all files are not using the correct line breaks:

![image](https://user-images.githubusercontent.com/5607939/110061410-37beea80-7d70-11eb-8d49-007df63408b4.png)

This is effecting many files, and I am not sure what the full impact will be on our Windows users. So I will push this change up to staging. Please take a look and let me know.

Sorry this took so long :+1:
